### ISSUE TITLE
CVE-2022-45688 in json-java transitive dependency for token-client library

### ISSUE THREAD

**Issue by Emuto** [SAP/cloud-security-services-integration-library#1186]
Hi, 

A security vulnerability (CVE-2022-45688) has been reported in json-java versions up to 20230227. 

Token-client 2.3.19 declares an affected version of json-java as a dependency, while token-client 3.0.0 declares the safe version 20230227.

Since we are using 2.3.19, my question would be - should we upgrade to 3.0.0 or will there be a new version of 2.3.x available with an updated dependency?  

Thank you and Best Regards,
Emanuil

**Comment by finkmanAtSap** [comment]
Hi Emanuil,

Token-client 2.3.19 should also use version 20230227.
We do not explicitly declare an old version in token-client:

```
<dependency>
	<groupId>org.json</groupId>
	<artifactId>json</artifactId>
</dependency>
```

Confirming it with `mvn depdenency:tree` I get

```
[INFO] --- maven-dependency-plugin:2.8:tree (default-cli) @ token-client ---
[INFO] com.sap.cloud.security.xsuaa:token-client:jar:2.13.9
...
[INFO] +- org.json:json:jar:20230227:compile
...
```

Where did you notice that it declares an affected version?

Also, when there was no fixed version available for the CVE, we had already added a test to assure our Wrapper class that uses org.json is not affected by the CVE which passes for the release:

https://github.com/SAP/cloud-security-services-integration-library/blob/main-2.x/env/src/test/java/com/sap/cloud/security/json/DefaultJsonObjectTest.java#L222

That being said, if you can afford the effort to upgrade to version 3.0.0, we recommend that of course! :)
There will however be additional releases for 2.x until the mentioned end of support plan in its README.
If I am mistaken and 2.13.9 somehow declares a dependency that is affected by the CVE, we would try to schedule a release to fix it.

Best,
Manuel

**Comment by Emuto** [comment]
Hi Manuel,

Thanks for the prompt reply! 

I can see that in your pom.xml for 2.13.9 indeed version 20230227 is set.

However, running dependency tree on our app produces the following:

 |  |  \- com.sap.cloud.security.xsuaa:token-client:jar:2.13.9:compile
[INFO] |  |     \- org.json:json:jar:20220320:compile

I'm not really sure why this happens as we don't override this version anywhere, it comes as a transitive dependency from token-client only..

**Comment by finkmanAtSap** [comment]
Hmm... That is odd.

I checked again and version 20230227 is specified in our repository's root pom in the dependency management section. As token-client uses that as parent, it automatically declares the dependency on 20230227, even if we don't mention any particular version in token-client's pom. Maybe it has something to do with this.

Do you have additional dependencies to this repository, other than token-client that are declared with a different version number? Maybe `mvn dependency:tree -Dverbose=true` can give more info?

**Comment by MarkR87** [comment]
Same problem on our side.
I think this is related to java-sap-vcap-services dependency which shows the mentioned vulnerability in the latest version.
https://mvnrepository.com/artifact/com.sap.cloud.environment.servicebinding/java-sap-vcap-services/0.5.2

**Comment by finkmanAtSap** [comment]
> Same problem on our side. I think this is related to java-sap-vcap-services dependency which shows the mentioned vulnerability in the latest version. https://mvnrepository.com/artifact/com.sap.cloud.environment.servicebinding/java-sap-vcap-services/0.5.2

Thanks for the hint. Looks like they have already pushed a fix in March but not released it yet: https://github.com/SAP/btp-environment-variable-access/commit/5aa1b688266c137a39ba77ab4747661d17d89dbf

We are declaring a dependency on it without version specification, so if they release, the new version should be used.

What I don't fully understand is why your app follows the version that java-sap-vcap-services specifies as the version specification in our parent pom should be higher in the maven dependency graph. For instance, if I view it from our POV, it looks like this:

![image](https://user-images.githubusercontent.com/123368068/235675554-7ba3aee7-1cd9-4f2d-9289-fb2a61e145af.png)

Does your app have a direct dependency on `java-sap-vcap-services` as well?

**Comment by Emuto** [comment]
No, we don't have it as a direct dependency. 

I have a hunch, however, that the core reason might be this: 

[INFO] |  |  +- com.sap.cloud.sdk.cloudplatform:security-scp-cf:jar:3.78.0:compile
[INFO] |  |  |  +- (com.sap.cloud.sdk.cloudplatform:cloudplatform-core:jar:3.78.0:compile - omitted for duplicate)
[INFO] |  |  |  +- (com.sap.cloud.sdk.cloudplatform:cloudplatform-core-scp-cf:jar:3.78.0:compile - omitted for duplicate)
[INFO] |  |  |  +- (com.sap.cloud.sdk.cloudplatform:servlet:jar:3.78.0:compile - omitted for duplicate)
[INFO] |  |  |  +- (com.sap.cloud.sdk.cloudplatform:resilience:jar:3.78.0:compile - omitted for duplicate)
[INFO] |  |  |  +- (com.sap.cloud.sdk.frameworks:resilience4j:jar:3.78.0:runtime - omitted for duplicate)
[INFO] |  |  |  +- (com.sap.cloud.sdk.cloudplatform:security:jar:3.78.0:compile - omitted for duplicate)
[INFO] |  |  |  +- (com.sap.cloud.security:java-security:jar:2.13.9:compile - version managed from 2.13.4; omitted for duplicate)
[INFO] |  |  |  +- (com.sap.cloud.security:env:jar:2.13.4:compile - version managed from 2.13.9; omitted for duplicate)

Following the dependency tree:

[INFO] +- com.sap.cloud.security:java-security:jar:2.13.9:compile
[INFO] |  +- com.sap.cloud.security:java-api:jar:2.13.4:compile
[INFO] |  +- com.sap.cloud.security:env:jar:2.13.4:compile (version managed from 2.13.9)
[INFO] |  |  +- (com.sap.cloud.security:java-api:jar:2.13.4:compile - version managed from 2.13.9; omitted for duplicate)
[INFO] |  |  +- com.sap.cloud.environment.servicebinding:java-sap-service-operator:jar:0.5.1:compile
[INFO] |  |  |  +- com.sap.cloud.environment.servicebinding.api:java-core-api:jar:0.5.1:compile
[INFO] |  |  |  |  \- (com.google.code.findbugs:jsr305:jar:3.0.2:compile - omitted for duplicate)
[INFO] |  |  |  +- (com.sap.cloud.environment.servicebinding.api:java-access-api:jar:0.5.1:compile - omitted for duplicate)
[INFO] |  |  |  +- (com.google.code.findbugs:jsr305:jar:3.0.2:compile - omitted for duplicate)
[INFO] |  |  |  +- (org.json:json:jar:20220320:compile - version managed from 20220924; omitted for duplicate)
[INFO] |  |  |  \- (org.slf4j:slf4j-api:jar:1.7.36:compile - version managed from 2.0.3; omitted for duplicate)
[INFO] |  |  +- com.sap.cloud.environment.servicebinding.api:java-consumption-api:jar:0.5.1:compile
[INFO] |  |  |  +- (com.sap.cloud.environment.servicebinding.api:java-core-api:jar:0.5.1:compile - omitted for duplicate)
[INFO] |  |  |  \- (com.google.code.findbugs:jsr305:jar:3.0.2:compile - omitted for duplicate)
[INFO] |  |  +- com.sap.cloud.environment.servicebinding.api:java-access-api:jar:0.5.1:compile
[INFO] |  |  |  +- (com.sap.cloud.environment.servicebinding.api:java-core-api:jar:0.5.1:compile - omitted for duplicate)
[INFO] |  |  |  +- (com.google.code.findbugs:jsr305:jar:3.0.2:compile - omitted for duplicate)
[INFO] |  |  |  \- (org.slf4j:slf4j-api:jar:1.7.36:compile - version managed from 2.0.3; omitted for duplicate)
[INFO] |  |  \- (org.json:json:jar:20220320:compile - version managed from 20220924; omitted for duplicate)

**Comment by finkmanAtSap** [comment]
Yes, thanks for this dependency tree. I agree, this is where the problems are coming from:

It seems the version of our `env` module (which has a dependency on both `org.json` and `java-sap-service-operator`) gets actively managed to version 2.13.4 by what seems to be Cloud SDK.
Unfortunately, `env` versions <= 2.13.6 are flagged (though to the best of our knowledge not at risk) with the mentioned CVE. So, even if `java-sap-service-operator` released a new version, I think the dependency on the old `org.json` version would remain due to the old `env` version.

The newest Cloud SDK version uses version 2.13.5 of our libs which is still too old.
I think you could try actively managing the version of our modules to 2.13.9 as a temporary solution until Cloud SDK adopts a newer release of ours.
